File: packages/core/workspace/operating-rules.test.ts
// @vitest-environment node
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";

import {
  DELIVERABLE_PLATFORMS,
  REVIEW_RULES,
  RULE_PLATFORMS,
  emptyOperatingRules,
  isDeliverable,
  operatingRulesToRequest,
  parseOperatingRules,
  readCadence,
  readObservation,
  templateNoteFor,
} from "./operating-rules";
import { isWebLink, readHomepage } from "./homepage";

// The controlled sets are the server's. Restating them here is unavoidable - a
// component has to switch on them - so they are compared against the Go source
// rather than against my memory of it, the same way 025's and 027's are.
const goSource = readFileSync(
  join(__dirname, "../../../server/internal/content/workspace-core/operating_rules.go"),
  "utf8",
);

function goSlice(name: string): string[] {
  const declaration = new RegExp(`var ${name} = \\[\\]string\\{([^}]*)\\}`, "s").exec(goSource);
  if (!declaration) throw new Error(`${name} not found in operating_rules.go`);
  return [...declaration[1]!.matchAll(/"([a-z_]+)"/g)].map((match) => match[1]!);
}

describe("the controlled sets match the Go source", () => {
  it("has ip-profile's eight channels, not review-delivery's four", () => {
    expect([...RULE_PLATFORMS]).toEqual(goSlice("Platforms"));
    expect(RULE_PLATFORMS).toHaveLength(8);
  });

  it("has exactly one review rule", () => {
    expect([...REVIEW_RULES]).toEqual(goSlice("ReviewRules"));
    expect(REVIEW_RULES).toHaveLength(1);
  });

  it("has no escape hatch in either set", () => {
    for (const forbidden of ["other", "misc", "custom", "team"]) {
      expect(RULE_PLATFORMS as readonly string[]).not.toContain(forbidden);
      expect(REVIEW_RULES as readonly string[]).not.toContain(forbidden);
    }
  });
});

describe("defaults", () => {
  it("fills the review rule and nothing else", () => {
    const rules = emptyOperatingRules();
    expect(rules.reviewRule).toBe("self");
    // No invented cadence and no invented window (FR-024).
    expect(rules.cadence).toEqual({});
    expect(rules.observation.default).toBeUndefined();
  });
});

describe("unset is not zero", () => {
  // The single silent-corruption risk on this card. A page that reads the
  // number and ignores `stored` cannot tell "nothing goes out on this channel"
  // from "nobody has decided", and those are different pages.
  it("tells a stored 0 apart from an absent channel", () => {
    const rules = emptyOperatingRules();
    rules.cadence.wechat_mp = 0;

    expect(readCadence(rules, "wechat_mp")).toEqual({ value: 0, stored: true });
    expect(readCadence(rules, "douyin")).toEqual({ value: 0, stored: false });
    expect(readCadence(rules, "wechat_mp")).not.toEqual(readCadence(rules, "douyin"));
  });

  it("does not turn a missing window into zero days", () => {
    const rules = emptyOperatingRules();
    expect(readObservation(rules, "xiaohongshu").source).toBe("none");

    rules.observation.default = 0;
    // A stored 0 IS a window: look the same day.
    expect(readObservation(rules, "xiaohongshu")).toEqual({ days: 0, source: "global" });
  });
});

describe("readObservation", () => {
  it("says where the window came from", () => {
    const rules = emptyOperatingRules();
    rules.observation.default = 14;
    rules.observation.byChannel.douyin = 7;

    expect(readObservation(rules, "douyin")).toEqual({ days: 7, source: "channel" });
    expect(readObservation(rules, "xiaohongshu")).toEqual({ days: 14, source: "global" });
  });
});

describe("parseOperatingRules", () => {
  it("reads what the server sends", () => {
    const rules = parseOperatingRules({
      cadence: { xiaohongshu: 3, wechat_mp: 0 },
      templates: { xiaohongshu: { note: "标题 20 字内" } },
      review_rule: "self",
      observation: { default: 14, by_channel: { douyin: 7 } },
    });
    expect(readCadence(rules, "wechat_mp")).toEqual({ value: 0, stored: true });
    expect(readCadence(rules, "douyin").stored).toBe(false);
    expect(templateNoteFor(rules, "xiaohongshu")).toBe("标题 20 字内");
    expect(readObservation(rules, "douyin")).toEqual({ days: 7, source: "channel" });
  });

  it("degrades a malformed response to defaults without inventing a window", () => {
    // Each of these is reachable from a real response: an older backend, a
    // null column, a field this build has not heard of.
    for (const wire of [null, undefined, "nope", 7, {}, { observation: null }, { cadence: null }]) {
      const rules = parseOperatingRules(wire);
      expect(rules.reviewRule).toBe("self");
      expect(rules.observation.default).toBeUndefined();
      expect(readCadence(rules, "xiaohongshu").stored).toBe(false);
    }
  });

  it("keeps a channel this build has not heard of rather than dropping it", () => {
    const rules = parseOperatingRules({ cadence: { newplatform: 2 }, observation: {} });
    expect(readCadence(rules, "newplatform")).toEqual({ value: 2, stored: true });
  });

  it("never turns an absent window into 0", () => {
    const rules = parseOperatingRules({ observation: { default: null } });
    expect(rules.observation.default).toBeUndefined();
    expect(rules.observation.default).not.toBe(0);
  });
});

describe("operatingRulesToRequest", () => {
  it("omits the window rather than sending null", () => {
    const body = operatingRulesToRequest(emptyOperatingRules());
    expect(body.observation).toEqual({ by_channel: {} });
    expect(JSON.stringify(body)).not.toContain("null");
  });

  it("sends a stored zero", () => {
    const rules = emptyOperatingRules();
    rules.cadence.wechat_mp = 0;
    rules.observation.default = 0;
    const body = operatingRulesToRequest(rules) as {
      cadence: Record<string, number>;
      observation: { default?: number };
    };
    expect(body.cadence.wechat_mp).toBe(0);
    expect(body.observation.default).toBe(0);
  });
});

describe("templateNoteFor", () => {
  // SOP 3.2: "模型上下文仅接收必要的渠道说明". Only the note, and a channel
  // with no template yields "" rather than undefined for a caller to handle.
  it("returns the note alone", () => {
    const rules = emptyOperatingRules();
    rules.cadence.xiaohongshu = 3;
    rules.templates.xiaohongshu = { note: "标题 20 字内" };
    rules.observation.default = 14;

    expect(templateNoteFor(rules, "xiaohongshu")).toBe("标题 20 字内");
    expect(templateNoteFor(rules, "zhihu")).toBe("");
  });
});

describe("isDeliverable", () => {
  // Eight channels can hold a template; four can be delivered to. Not a bug -
  // FR-012a - but the page has to say it, so there is a function to ask.
  it("separates the four from the eight", () => {
    expect(DELIVERABLE_PLATFORMS).toHaveLength(4);
    expect(isDeliverable("xiaohongshu")).toBe(true);
    expect(isDeliverable("zhihu")).toBe(false);
    for (const platform of DELIVERABLE_PLATFORMS) {
      expect(RULE_PLATFORMS as readonly string[]).toContain(platform);
    }
  });
});

describe("homepage", () => {
  it("accepts a web link", () => {
    expect(isWebLink("https://www.xiaohongshu.com/user/profile/x")).toBe(true);
    expect(isWebLink("http://example.com/me")).toBe(true);
  });

  it("refuses anything that is not a page a person can open", () => {
    for (const link of ["", "javascript:alert(1)", "file:///etc/passwd", "data:text/html,x", "www.example.com"]) {
      expect(isWebLink(link)).toBe(false);
    }
  });

  it("tells a cleared link apart from one nobody entered", () => {
    expect(readHomepage({ "loretide.homepage": "" })).toEqual({ link: "", stored: true });
    expect(readHomepage({})).toEqual({ link: "", stored: false });
    expect(readHomepage(null)).toEqual({ link: "", stored: false });
    expect(readHomepage({ "loretide.homepage": 7 })).toEqual({ link: "", stored: false });
  });

  it("leaves the account's other settings out of its business", () => {
    // LT-014's scope lives on the same blob. Reading a homepage must not need
    // to know that, and must not disturb it.
    const settings = { "loretide.scope": "all", "loretide.homepage": "https://example.com" };
    expect(readHomepage(settings).link).toBe("https://example.com");
    expect(settings["loretide.scope"]).toBe("all");
  });
});

File: server/internal/content/workspace-core/operating_rules_test.go
package workspacecore

import (
	"encoding/json"
	"errors"
	"strings"
	"testing"
)

// The rules a page asks about before it enables a button, and the one thing
// this card cannot get wrong: a number nobody entered is not zero.

func days(value int64) *int64 { return &value }

func TestTheControlledSetsAreExactlyWhatTheSOPNames(t *testing.T) {
	if len(Platforms) != 8 {
		t.Fatalf("Platforms has %d entries, want 8", len(Platforms))
	}
	// SOP 3.2 says "个人默认自己审核。后续团队模式可要求不同成员复核。" - the
	// second sentence is later work, and a second value here today would
	// promise something nothing implements.
	if len(ReviewRules) != 1 || ReviewRules[0] != ReviewRuleSelf {
		t.Fatalf("ReviewRules = %v, want exactly [self]", ReviewRules)
	}
	for _, forbidden := range []string{"other", "misc", "custom", "team"} {
		for _, platform := range Platforms {
			if platform == forbidden {
				t.Errorf("Platforms contains an escape hatch %q", forbidden)
			}
		}
		for _, rule := range ReviewRules {
			if rule == forbidden {
				t.Errorf("ReviewRules contains an escape hatch %q", forbidden)
			}
		}
	}
}

func TestDefaultsFillWhatTheSOPGivesAndNothingElse(t *testing.T) {
	rules := DefaultRules()
	if rules.ReviewRule != ReviewRuleSelf {
		t.Errorf("review rule default = %q, want %q", rules.ReviewRule, ReviewRuleSelf)
	}
	// No default cadence and no default observation window. Inventing either
	// would be a rule the SOP never stated (FR-024).
	if len(rules.Cadence) != 0 {
		t.Errorf("a brand that set nothing has a cadence: %v", rules.Cadence)
	}
	if rules.Observation.Default != nil {
		t.Errorf("a brand that set nothing has an observation window: %v", *rules.Observation.Default)
	}
}

// The single silent-corruption risk on this card, and the reason every reader
// returns two values. 019 hit the same thing on a boolean.
func TestUnsetIsNotZero(t *testing.T) {
	rules := DefaultRules()
	rules.Cadence["xiaohongshu"] = 0

	zero, storedZero := ReadCadence(rules, "xiaohongshu")
	_, storedAbsent := ReadCadence(rules, "douyin")

	if zero != 0 || !storedZero {
		t.Errorf("a stored 0 read back as (%d, %v), want (0, true)", zero, storedZero)
	}
	if storedAbsent {
		t.Error("a channel nobody set reads as stored; 'nothing goes out' and 'nobody decided' are different")
	}
	if storedZero == storedAbsent {
		t.Error("a stored 0 is indistinguishable from an absent value")
	}
}

func TestObservationSourceIsThreeWayNotBoolean(t *testing.T) {
	rules := DefaultRules()
	rules.Observation.Default = days(14)
	rules.Observation.ByChannel["douyin"] = 7

	for _, tc := range []struct {
		name     string
		platform string
		want     int64
		source   ObservationSource
	}{
		{"the channel set its own", "douyin", 7, ObservationFromChannel},
		{"the channel falls back to the brand", "xiaohongshu", 14, ObservationFromGlobal},
	} {
		t.Run(tc.name, func(t *testing.T) {
			value, source := ReadObservation(rules, tc.platform)
			if value != tc.want || source != tc.source {
				t.Errorf("= (%d, %q), want (%d, %q)", value, source, tc.want, tc.source)
			}
		})
	}

	// Nothing set anywhere: not 0-from-global, not a silent zero.
	bare := DefaultRules()
	if _, source := ReadObservation(bare, "xiaohongshu"); source != ObservationUnset {
		t.Errorf("source with nothing set = %q, want %q", source, ObservationUnset)
	}
}

func TestAStoredZeroObservationIsARealWindow(t *testing.T) {
	rules := DefaultRules()
	rules.Observation.Default = days(0)
	value, source := ReadObservation(rules, "xiaohongshu")
	if value != 0 || source != ObservationFromGlobal {
		t.Errorf("= (%d, %q), want (0, global) - 'look the same day' is a decision", value, source)
	}
}

func TestValidateNamesTheFieldThatIsWrong(t *testing.T) {
	for _, tc := range []struct {
		name  string
		rules func(Rules) Rules
		field string
	}{
		{"a negative cadence", func(r Rules) Rules {
			r.Cadence["xiaohongshu"] = -1
			return r
		}, "cadence.xiaohongshu"},
		{"a channel outside the eight", func(r Rules) Rules {
			r.Cadence["twitter"] = 3
			return r
		}, "cadence"},
		{"a review rule that is not self", func(r Rules) Rules {
			r.ReviewRule = "team"
			return r
		}, "review_rule"},
		{"a negative global window", func(r Rules) Rules {
			r.Observation.Default = days(-3)
			return r
		}, "observation.default"},
		{"a negative channel window", func(r Rules) Rules {
			r.Observation.ByChannel["douyin"] = -1
			return r
		}, "observation.by_channel.douyin"},
		{"a template on an unknown channel", func(r Rules) Rules {
			r.Templates["twitter"] = ChannelTemplate{Note: "x"}
			return r
		}, "templates"},
		{"a note past the rune limit", func(r Rules) Rules {
			r.Templates["zhihu"] = ChannelTemplate{Note: strings.Repeat("字", MaxNoteRunes+1)}
			return r
		}, "templates.zhihu"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			err := ValidateRules(tc.rules(DefaultRules()))
			var field FieldError
			if !errors.As(err, &field) {
				t.Fatalf("err = %v, want a FieldError", err)
			}
			if field.Field != tc.field {
				t.Errorf("named %q, want %q", field.Field, tc.field)
			}
			if !errors.Is(err, ErrInvalid) {
				t.Error("does not unwrap to ErrInvalid, so the handler cannot map it to 400")
			}
		})
	}
}

func TestValidateAcceptsWhatTheSOPAllows(t *testing.T) {
	rules := DefaultRules()
	// 0 is legal on both: "nothing goes out this week" and "look the same day".
	rules.Cadence["xiaohongshu"] = 0
	rules.Observation.Default = days(0)
	// A note exactly at the limit, counted in runes.
	rules.Templates["zhihu"] = ChannelTemplate{Note: strings.Repeat("字", MaxNoteRunes)}
	// A template on a channel delivery cannot reach today is fine (FR-012a).
	rules.Templates["bilibili"] = ChannelTemplate{Note: "长视频，封面 16:9"}
	if err := ValidateRules(rules); err != nil {
		t.Fatalf("ValidateRules = %v, want nil", err)
	}
}

// FR-012a: eight channels can hold a template; four of them cannot be
// delivered to yet. That is a fact the page has to say out loud, not something
// storage should refuse.
func TestATemplateMayLiveOnAChannelDeliveryCannotReach(t *testing.T) {
	undeliverable := []string{"bilibili", "zhihu", "weibo", "kuaishou"}
	rules := DefaultRules()
	for _, platform := range undeliverable {
		rules.Templates[platform] = ChannelTemplate{Note: "写法"}
	}
	if err := ValidateRules(rules); err != nil {
		t.Fatalf("refused a template on a channel that cannot be delivered to: %v", err)
	}
	for _, platform := range undeliverable {
		if TemplateNoteFor(rules, platform) == "" {
			t.Errorf("%s lost its note", platform)
		}
	}
}

func TestRulesSurviveARoundTripThroughTheSettingsBlob(t *testing.T) {
	rules := DefaultRules()
	rules.Cadence["xiaohongshu"] = 3
	rules.Cadence["wechat_mp"] = 0
	rules.Templates["xiaohongshu"] = ChannelTemplate{Note: "标题 20 字内"}
	rules.Observation.Default = days(14)
	rules.Observation.ByChannel["douyin"] = 7

	encoded, err := json.Marshal(rules)
	if err != nil {
		t.Fatal(err)
	}
	var stored any
	if err := json.Unmarshal(encoded, &stored); err != nil {
		t.Fatal(err)
	}
	read := RulesFrom(map[string]any{OperatingRulesKey: stored})

	if value, ok := ReadCadence(read, "wechat_mp"); value != 0 || !ok {
		t.Errorf("a stored 0 did not survive the round trip: (%d, %v)", value, ok)
	}
	if _, ok := ReadCadence(read, "douyin"); ok {
		t.Error("an absent cadence came back as stored")
	}
	if value, source := ReadObservation(read, "douyin"); value != 7 || source != ObservationFromChannel {
		t.Errorf("channel window = (%d, %q)", value, source)
	}
	if TemplateNoteFor(read, "xiaohongshu") != "标题 20 字内" {
		t.Error("the note did not survive")
	}
}

func TestAnUnreadableBlobDegradesToDefaults(t *testing.T) {
	// Every one of these is reachable from a real row: a brand that predates
	// this card, a null settings column, and a blob written by something that
	// did not know the key.
	for _, settings := range []any{nil, "not an object", []any{1, 2}, map[string]any{}, map[string]any{OperatingRulesKey: 7}} {
		rules := RulesFrom(settings)
		if rules.ReviewRule != ReviewRuleSelf {
			t.Errorf("settings %v degraded to review rule %q", settings, rules.ReviewRule)
		}
		if rules.Observation.Default != nil {
			t.Errorf("settings %v invented an observation window", settings)
		}
	}
}

func TestHomepageOnlyTakesWebLinks(t *testing.T) {
	for _, link := range []string{
		"https://www.xiaohongshu.com/user/profile/x",
		"http://example.com/me",
		"",
	} {
		if err := ValidateHomepage(link); err != nil {
			t.Errorf("ValidateHomepage(%q) = %v, want nil", link, err)
		}
	}
	// Not a fetch concern - nothing here follows the link - but a person will
	// click it, and these are what a click should never do.
	for _, link := range []string{
		"javascript:alert(1)",
		"file:///etc/passwd",
		"data:text/html,<script>",
		"ftp://example.com",
		"www.example.com",
	} {
		var field FieldError
		err := ValidateHomepage(link)
		if !errors.As(err, &field) || field.Field != "homepage" {
			t.Errorf("ValidateHomepage(%q) = %v, want a homepage FieldError", link, err)
		}
	}
}

func TestHomepageReadDistinguishesClearedFromNeverSet(t *testing.T) {
	cleared, storedCleared := HomepageFrom(map[string]any{HomepageKey: ""})
	_, storedNever := HomepageFrom(map[string]any{})
	if cleared != "" || !storedCleared {
		t.Errorf("a cleared link read as (%q, %v), want (\"\", true)", cleared, storedCleared)
	}
	if storedNever {
		t.Error("a link nobody entered reads as stored")
	}
}

// SOP 3.2: "模型上下文仅接收必要的渠道说明". The note, and nothing standing
// next to it.
func TestOnlyTheChannelNoteCanReachAModel(t *testing.T) {
	rules := DefaultRules()
	rules.Cadence["xiaohongshu"] = 3
	rules.Templates["xiaohongshu"] = ChannelTemplate{Note: "标题 20 字内"}
	rules.Observation.Default = days(14)
	rules.ReviewRule = ReviewRuleSelf

	note := TemplateNoteFor(rules, "xiaohongshu")
	if note != "标题 20 字内" {
		t.Fatalf("note = %q", note)
	}
	for _, leaked := range []string{"3", "14", ReviewRuleSelf, "cadence", "observation"} {
		if strings.Contains(note, leaked) {
			t.Errorf("the model context carries %q, which is not a channel note", leaked)
		}
	}
}
